modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/seedlist/ranges.scala

```diff
@@ -0,0 +1,67 @@
+package org.tessellation.node.shared.domain.seedlist
+
+import cats.data.NonEmptyList
+import cats.syntax.all._
+
+import scala.util.control.NoStackTrace
+
+import org.tessellation.node.shared.domain.seedlist.Range.ValidationResult
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.types.numeric.NonNegLong
+
+case class InvalidInput(str: String) extends NoStackTrace {
+  override def getMessage = s"Invalid input: $str"
+}
+
+case object NoValidRanges extends NoStackTrace
+
+object ranges {
+
+  private val delimiter = ";"
+
+  def decode(input: String): ValidationResult[NonEmptyList[Range]] =
+    input
+      .split(delimiter)
+      .toList
+      .traverse(Range.decode)
+      .andThen {
+        validate(_)
+          .leftMap(_ => InvalidInput(input))
+          .toValidatedNel
+      }
+      .map(mergeOverlaps)
+
+  private def mergeOverlaps(ranges: NonEmptyList[Range]): NonEmptyList[Range] = {
+    val merged = ranges
+      .sortBy(_.start.value)
+      .foldLeft(List.empty[Range]) {
+        case (prev :: tail, curr) if curr.start - prev.end <= 1 =>
+          val merged = Range(
+            prev.start,
+            NonNegLong.unsafeFrom(
+              Math.max(prev.end, curr.end)
+            )
+          )
+
+          merged :: tail
+        case (acc, curr) =>
+          curr :: acc
+      }
+      .reverse
+
+    NonEmptyList(merged.head, merged.tail)
+  }
+
+  private def validate(ranges: List[Range]): Either[Throwable, NonEmptyList[Range]] = {
+    val filtered = ranges.filter(r => r.start <= r.end)
+
+    Either
+      .cond(
+        filtered.nonEmpty,
+        NonEmptyList(filtered.head, filtered.tail),
+        NoValidRanges
+      )
+  }
```

Comment:
You can use `toNel` and `toRight`, so you don't need to use unsafe `.head` operation.

```scala
  private def validate(ranges: List[Range]): Either[Throwable, NonEmptyList[Range]] =
    ranges
      .filter(r => r.start <= r.end)
      .toNel
      .toRight(NoValidRanges)
  ```

Reply:
Changed to:

```scala
  def decode(input: String): ValidationResult[NonEmptyList[Range]] =
    input
      .split(delimiter)
      .toList
      .traverse(Range.decode)
      .andThen {
        _.filter(r => r.start <= r.end).toNel
          .toRight(InvalidInput(input))
          .toValidatedNel
      }
      .map(mergeOverlaps)
```